### DISCUSSION TITLE
Getting Started with Seeed Studio XIAO MG24 | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#2030] [Q&A]
# Getting Started with Seeed Studio XIAO MG24 | Seeed Studio Wiki

https://wiki.seeedstudio.com/xiao_mg24_getting_started/

**Comment by msfujino** [comment]
XIAO_MG24 is designed to convert the maximum 4.2V battery voltage to 2.1V by dividing by 1/2 and then AD converting it.
In BSP2.2.0, the default setting is VREF=1.2V and gain=1, and as it is, AD conversion is possible only up to 1.2V.
If the gain is set to 0.5, it can convert up to 2.4V, but the gain cannot be changed because no function is provided. 
Is there any way to do this?

**Reply by nfs0619** [reply]
Thank you for your inquiry! Regarding your question, the default VREF is 3.3V, and the setting of VREF=1.2V with gain=1 is intended for optimizing the ADC conversion for the microphone (MIC). If you need to support higher voltage ranges for ADC conversion (e.g., up to 2.4V), it’s recommended to use an external 3.3V reference voltage. This will allow the ADC to convert higher voltages, which is necessary for measuring battery voltage.

You can refer to the following links for details on how to adjust VREF and related settings:

[Modify reference voltage](https://github.com/SiliconLabs/arduino/blob/main/cores/silabs/wiring_analog.cpp#L45)
[Possible reference voltage values](https://github.com/SiliconLabs/arduino/blob/main/cores/silabs/adc.h#L39)
Hope this helps!

**Reply by msfujino** [reply]
If Vref=3.3V, measurement will not be possible if the battery voltage falls below 3.3V.
A function that can be set gain=0.5 is required.

**Reply by nfs0619** [reply]
Thank you for your follow-up! To better assist you with this issue, we recommend reaching out to our technical support team. They will be able to provide more detailed assistance based on your specific needs. You can contact them via the following email:

**[techsupport@seeed.io](mailto:techsupport@seeed.io)**

Hope this helps resolve your issue!

**Reply by msfujino** [reply]
BSP 2.3.0 was released yesterday, but this issue was not resolved.
This is a real shame, as it is an essential feature for running the MG24 on battery power.
techsupport does not seem to recognize the importance of this issue.

**Comment by msfujino** [comment]
The OS is Linux Mint 21.3(Ubunts 22.04).
I am trying to upload a Blink example using ArduinoIDE 2.3.4 or 1.8.19, but I get the following error and cannot upload it.

Is there a way to deal with this?


/home/fujino/.arduino15/packages/SiliconLabs/tools/gcc-arm-none-eabi/12.2.rel1/bin/arm-none-eabi-size -A /home/fujino/.cache/arduino/sketches/E292C634CBD217E78FA7B79A76DC1D5A/Blink.ino.elf
Sketch uses 235492 bytes (14%) of program storage space. Maximum is 1572864 bytes.
Global variables use 31284 bytes (11%) of dynamic memory, leaving 230860 bytes for local variables. Maximum is 262144 bytes.
"/home/fujino/.arduino15/packages/SiliconLabs/tools/openocd/0.12.0-arduino1-static/bin/openocd" -d2 -s "/home/fujino/.arduino15/packages/SiliconLabs/tools/openocd/0.12.0-arduino1-static/share/openocd/scripts/" -f interface/cmsis-dap.cfg -f target/efm32s2_g23.cfg -c "init; reset_config srst_nogate; reset halt; program {/home/fujino/.cache/arduino/sketches/E292C634CBD217E78FA7B79A76DC1D5A/Blink.ino.hex}; reset; exit"
Open On-Chip Debugger 0.12.0+dev-01514-g21fa2de70 (2024-02-07-19:19)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 2
Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
efm32s2_dci_read_se_status
Error: unable to open CMSIS-DAP device 0x2886:0x62
Error: unable to find a matching CMSIS-DAP device

Failed uploading: uploading error: exit status 1

**Reply by msfujino** [reply]
BSP 2.3.0 has been released and I was hoping for an improvement, but unfortunately I still cannot upload on Linux.

**Reply by msfujino** [reply]
The following link provides a solution.

https://forum.arduino.cc/t/cannot-upload-to-xiao-mg24-on-linux-mint/1365765

**Reply by nfs0619** [reply]
Thank you for sharing!

**Comment by msfujino** [comment]
I am having a problem with one of my 3 MG24's, 2 have no problem.

I am trying to upload a Blink Example on Windows 11 ArduinoIDE 2.3.4, but I get the following error and cannot upload it I have tried on 3 different Windows PCs and the problem persists.
Is there a workaround?


Open On-Chip Debugger 0.12.0+dev-01514-g21fa2de70 (2024-02-07-19:18)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 2
Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
efm32s2_dci_read_se_status
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2886:0x0062, serial=B5B16EE3
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Serial# = B5B16EE3
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 1 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Error: Error connecting DP: cannot read IDR


Failed uploading: uploading error: exit status 1

**Reply by nfs0619** [reply]
Thank you for your feedback. We have noted the issue you're experiencing, and the development team is actively investigating it. They will reach out to you with an update once progress is made. We appreciate your patience and understanding!

**Reply by msfujino** [reply]
After uploading the DeepSleep Example from the MG24 Wiki, I am unable to upload any other sketches.
I think MG24 is sleeping and not responding to uploads.

**Reply by msfujino** [reply]
My experimental countermeasure is to have a period of time when MG24 is not in DeepSleep (e.g., delay(10000);) and reset it at the perfect time so that the IDE will start uploading during this period.
However, this method requires an unnecessary delay(10000); for each wake-up.

**Reply by waldow** [reply]
I also bricked some boards playing with deep sleep . I hope I can unbrick them using ST-LINK ?

**Reply by msfujino** [reply]
It depends on what sketches are uploaded to MG24 now.
Reset MG24 so that the IDE will start uploading new sketche during periods when MG24 is not DeepSlepp (e.g. long delay()). You may have to try many times. If you are lucky, it will recover.
For reference, you can also look at the forum.
https://forum.seeedstudio.com/t/deepsleep-bricks-xiao-mg24/284889

**Comment by msfujino** [comment]
The pin definition of the flash memory in “Demo3 : Deep Sleep Mode with flash and wake-up” is wrong.

#define CS_PIN PA6      ----> OK
#define CLK_PIN PA3    ----> PA0 
#define MOSI_PIN PA5 ----> PB0
#define MISO_PIN PA4 ----> PB1

**Reply by msfujino** [reply]
The pin definition of SW in “Demo3 : Deep Sleep Mode with flash and wake-up” is wrong.  

//SW
  pinMode(PD3, OUTPUT);
  pinMode(PB5, OUTPUT);
  pinMode(PB1, OUTPUT);                ---->    PD5
  pinMode(PB0, OUTPUT);                ---->    PC8
  pinMode(PA6, OUTPUT);
  digitalWrite(PD3, LOW); //VBAT
  digitalWrite(PB5, LOW); //RF_SW
  digitalWrite(PB1, LOW); //IMU        ---->    PD5
  digitalWrite(PB0, LOW); //MIC        ---->    PC8
  digitalWrite(PA6, HIGH);  //FLASH

**Reply by waldow** [reply]
Thank you  @msfujino. I did not got the claimed 1.95μA was getting 7μA . Your tweaks fixed that.

**Reply by msfujino** [reply]
For reference, you can also look at the forum.
https://forum.seeedstudio.com/t/light-deep-sleep-current-comparison-of-xiao-mg24-and-xiao-nrf52840/284910

**Comment by eerke2021** [comment]
I don’t believe Thread is available in Arduino: which software do we use?
Ditto the Machine learning: is this in Arduino or a different software?

**Reply by limengdu** [reply]
The chip supports it, but the chip originals are in development and need to wait.

**Reply by eerke2021** [reply]
If I use the Silicon Labs IDE (I think it is called "simplicity") can I connect it to a thread network? 

The ML aspect I am not that interested in but I'd like to add it to a thread network.

**Reply by limengdu** [reply]
Sorry, we don't have any relevant experience with Thread at this time.

**Comment by bindaz** [comment]
Hi can you write a code example for Xiao MG24 sleep mode and wake up by the IMU motion?

**Reply by limengdu** [reply]
Thank you for your feedback, we will arrange for our colleagues to be scheduled to do such a procedure.

**Comment by rajkmrhenry-dot** [comment]
Is the battery charging current 200mA, can i control it?

**Reply by bleaaach** [reply]
Hi, the battery charging current of XIAO MG24 is indeed 200mA. This is a fixed setting of the onboard charging management chip and cannot be adjusted through software or external control.

**Comment by jimthomas470-max** [comment]
Using Arduino IDE,  need advice...
Trying to tweak the BLE TX power and getting status error 0x100C  SL_STATUS_BT_CTRL_COMMAND_DISALLOWED

Code below placed after boot event received ...  case sl_bt_evt_system_boot_id:
Also tried in Setup loop, same result

  sl_bt_system_halt(1);  // suspend radio while changing power
  status = sl_bt_system_set_tx_power(MIN_TX_POWER_DBM, MAX_TX_POWER_DBM, &set_min, &set_max);
  sl_bt_system_halt(0);  // restart radio after changing power

**Reply by bleaaach** [reply]
Thank you for your feedback. Regarding the issue of encountering error 0x100C when adjusting the BLE transmit power on XIAO MG24, we recommend referring to the official documentation for proper usage, as this part is not currently covered in our wiki.
https://docs.silabs.com/bluetooth/latest/bluetooth-stack-api/sl-bt-system
https://docs.silabs.com/bluetooth/5.0/bluetooth-general-system-and-performance/bluetooth-tx-power-settings
